### PR TITLE
Implemented ReflectionFunctionAbstract::hasReturnType and getReturnType

### LINKED ISSUE #419: Reflection has unimplemented methods required by PHPUnit and Mockery
The following methods need to be implemented for PHPUnit and Mockery:
- [x] ReflectionClass::implementsInterface
- [x] ReflectionParameter::isVariadic 
- [x] ReflectionMethod::hasReturnType
- [x] ReflectionMethod::getReturnType
- [x] ReflectionMethod::isProtected
- [x] ReflectionMethod::isPrivate

### PR DESCRIPTION
**Pull request by calvinbaart**
Fixes #419

### CODE REVIEW

**Comment by jakubmisek** [comment]
I was thinking if we can implement it just using .NET reflection without an explicit `[ReturnTypeAttribute]`

In this way it won't work for regular C# methods.

**Comment by calvinbaart** [comment]
The problems I have encountered with that are that sometimes the optimizer will give the class a return type. For example in the following sample:
```php
function func1()
{
    return 10;
}
```
PHP will return false on hasReturnType but the compiled version will be
```csharp
long func1()
{
    return 10;
}
```

File: src/Peachpie.Library/Reflection/ReflectionType.cs [hunk 1]
```diff
@@ -11,19 +11,24 @@ namespace Pchp.Library.Reflection
     [PhpType(PhpTypeAttribute.InheritName), PhpExtension(ReflectionUtils.ExtensionName)]
     public class ReflectionType
     {
-        public virtual bool allowsNull() { throw new NotImplementedException(); }
+        public string Name { get; set; }
+        public bool AllowsNull { get; set; }
+        public bool BuiltIn { get; set; }
```
Comment (jakubmisek) [inline src/Peachpie.Library/Reflection/ReflectionType.cs]: The class should not have properties `Name`, `AllowsNull` and `BuiltIn` visible (https://www.php.net/manual/en/class.reflectiontype.php)

They should be marked `internal` or `internal protected`.

File: src/Peachpie.Runtime/Attributes.cs [hunk 2]
```diff
@@ -287,6 +287,26 @@ public sealed class NotNullAttribute : Attribute
...
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method)]
+    public sealed class ReturnTypeAttribute : Attribute
+    {
+        /// <summary>
+        /// The type containing exported functions.
+        /// </summary>
+        public string SyntaxType { get; private set; }
+
+        public bool AllowsNull { get; private set; }
```
Comment (jakubmisek) [inline src/Peachpie.Runtime/Attributes.cs]: We already have attribute `[return: NotNull]` which should be emitted if the return type cannot be `null`.

File: src/Peachpie.Runtime/Attributes.cs [hunk 3]
```diff
@@ -287,6 +287,26 @@ public sealed class NotNullAttribute : Attribute
 
     }
 
+    /// <summary>
+    /// Compiler generated attribute denoting that the method returns a specific type
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method)]
+    public sealed class ReturnTypeAttribute : Attribute
```
Comment (jakubmisek) [inline src/Peachpie.Runtime/Attributes.cs]: I think this is redundant - we could rely on .NET type. Annotating every method with an attribute and using too much strings may not worth it.

**Comment by jakubmisek** [comment]
Yes, it would annotate return types for private, static and global functions, maybe more in future. But is it problem for PHP apps?

**Comment by jakubmisek** [comment]
Also nullable value types would be seen as `mixed` (because now they compile as `PhpValue` instead of `Nullable<long>` for example. :)

**Comment by jakubmisek** [comment]
proposing CLR based ReflectionNamedType
https://github.com/peachpiecompiler/peachpie/commit/13f66d5bef0764cca77ba8bcf952064c39ed41db

This will work in general while we will fix/implement additional cases by adding custom attributes, when needed.

**Pull request closed without merging**